### ISSUE TITLE
"numRequestsAllowed" property in rateLimiter not set after upgrade from 3.x to 4.x

### ISSUE THREAD

**Issue by jacobgkau** [RocketChat/Rocket.Chat#26089]
### Description:

The RocketCat bot sent out a message today and placed a banner at the top of my instance, running the most up-to-date version of the 3.x Snap track, stating a security vulnerability existed and I needed to upgrade to version 4.4 or above.

I upgraded to version 4.x with `snap switch rocketchat-server --channel=4.x/stable` followed by `snap refresh`.

First, the server failed to start because of MongoDB errors like these:

```
Error creating index: users -> { bio: 1 } { sparse: 1 } MongoError: Index with name: bio_1 already exists with different options
Error creating indexes for integration_history MongoError: Index with name: _updatedAt_1 already exists with different options
```

Following the advice in https://github.com/RocketChat/Rocket.Chat/issues/26046#issuecomment-1170177789, I was able to resolve these errors by dropping indexes from the database. However, I'm now unable to start with this error:

```
2022-07-01T03:02:07Z rocketchat-server.rocketchat-server[405028]: /snap/rocketchat-server/1520/programs/server/node_modules/fibers/future.js:280
2022-07-01T03:02:07Z rocketchat-server.rocketchat-server[405028]:                                               throw(ex);
2022-07-01T03:02:07Z rocketchat-server.rocketchat-server[405028]:                                               ^
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]: errorClass [Error]: [You must set "numRequestsAllowed" property in rateLimiter for REST API endpoint]
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at APIClass.addRateLimiterRuleForRoutes (app/api/server/api.js:263:10)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at APIClass.addRoute (app/api/server/api.js:336:9)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at module (app/api/server/v1/users.js:368:8)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at fileEvaluate (packages/modules-runtime.js:336:7)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.require (packages/modules-runtime.js:238:14)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.moduleLink [as link] (/snap/rocketchat-server/1520/programs/server/npm/node_modules/meteor/modules/node_modules/@meteorjs/reify/lib/runtime/index.js:52:22)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at module (app/api/server/index.ts:1:21)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at fileEvaluate (packages/modules-runtime.js:336:7)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.require (packages/modules-runtime.js:238:14)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.moduleLink [as link] (/snap/rocketchat-server/1520/programs/server/npm/node_modules/meteor/modules/node_modules/@meteorjs/reify/lib/runtime/index.js:52:22)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at module (app/apps/server/communication/rest.js:1:181)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at fileEvaluate (packages/modules-runtime.js:336:7)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.require (packages/modules-runtime.js:238:14)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.moduleLink [as link] (/snap/rocketchat-server/1520/programs/server/npm/node_modules/meteor/modules/node_modules/@meteorjs/reify/lib/runtime/index.js:52:22)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at module (app/apps/server/communication/index.js:1:321)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at fileEvaluate (packages/modules-runtime.js:336:7)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.require (packages/modules-runtime.js:238:14)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.moduleLink [as link] (/snap/rocketchat-server/1520/programs/server/npm/node_modules/meteor/modules/node_modules/@meteorjs/reify/lib/runtime/index.js:52:22)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at module (app/apps/server/orchestrator.js:1:1126)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at fileEvaluate (packages/modules-runtime.js:336:7)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.require (packages/modules-runtime.js:238:14)
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:     at Module.moduleLink [as link] (/snap/rocketchat-server/1520/programs/server/npm/node_modules/meteor/modules/node_modules/@meteorjs/reify/lib/runtime/index.js:52:22) {
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:   isClientSafe: true,
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:   error: 'You must set "numRequestsAllowed" property in rateLimiter for REST API endpoint',
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:   reason: undefined,
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:   details: undefined,
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]:   errorType: 'Meteor.Error'
2022-07-01T03:02:08Z rocketchat-server.rocketchat-server[405028]: }
```

### Steps to reproduce:

Covered above.

### Expected behavior:

The server should start.

### Actual behavior:

Server fails to start with the above error message.

### Server Setup Information:

- Version of Rocket.Chat Server: 4.8.1
- Operating System: Ubuntu 20.04.4
- Deployment Method: Snap
- Number of Running Instances: 1 (currently 0)
- DB Replicaset Oplog: ???
- NodeJS Version: ???
- MongoDB Version: 4.2.17

### Client Setup Information

- Desktop App or Browser Version: Desktop app version 3.8.6
- Operating System: Arch Linux, Android, Windows, iOS

### Additional context

I do not know where "numRequestsAllowed" is supposed to be set, especially for a Snap installation. The only information I could find about this was the PR adding the error message: https://github.com/RocketChat/Rocket.Chat/pull/11251 

### Relevant logs:

The instructions to gather logs from the admin panel in your template's link do not work when the server fails to start.

**Comment by jacobgkau** [comment]
Upon reviewing the logs closer, this "numRequestsAllowed" error was showing up even before I resolved the other two error messages regarding the MongoDB indexes, so dropping the indexes is not what caused this.

**Comment by jacobgkau** [comment]
Very difficult to find info about this error message with web searches, but I just found #26058 that sounds like it may fix this when it's released.

**Comment by anois** [comment]
@jacobgkau I fixed this by adding `Rate_Limiter_Limit_RegisterUser` to `rocketchat_settings`.

```javascript
db.rocketchat_settings.insert({ 
  "_id" : "Rate_Limiter_Limit_RegisterUser",
  "createdAt" : ISODate("2022-07-01T05:32:48.791Z"),
  "value" : 1,
  "packageValue" : 1,
  "valueSource" : "packageValue",
  "secret" : false,
  "enterprise" : false,
  "i18nDescription" : "Rate_Limiter_Limit_RegisterUser_Description",
  "autocomplete" : true,
  "sorter" : 0,
  "ts" : ISODate("2022-07-01T05:32:48.791Z"),
  "type" : "int",
  "group" : "Rate Limiter",
  "section" : "Feature_Limiting",
  "enableQuery" : "{\"_id\":\"API_Enable_Rate_Limiter\",\"value\":true}",
  "i18nLabel" : "Rate_Limiter_Limit_RegisterUser",
  "hidden" : false,
  "blocked" : false,
  "requiredOnWizard" : false,
  "env" : false,
  "public" : false,
  "_updatedAt" : ISODate("2022-07-01T05:38:12.246Z")
})
```

**Comment by tuxfamily** [comment]
I'm in the same situation (upgrading from 3.18 to 4.8 because of cve-2022-32211).
I also encoutered the index issue, plus some AppArmor rules issues.
Now I'm stuck with this error. I tried to run the mongo query from @anois but it does not change anything.

**Comment by tuxfamily** [comment]
I played a bit with the settings stored in mongodb and these queries seem to have "worked".

```go
db.getSiblingDB
("parties").getCollection("rocketchat_settings").updateOne({_id: "API_Enable_Rate_Limiter"}, {"$set": {value: false}});

db.getSiblingDB
("parties").getCollection("rocketchat_settings").updateOne({_id: "API_Enable_Rate_Limiter_Dev"}, {"$set": {value: false}});
```

This leaded me other errors but it's not the point of this issue so... good luck :)

**Comment by jacobgkau** [comment]
The workaround in https://github.com/RocketChat/Rocket.Chat/issues/26089#issuecomment-1172211607 worked for me, and my server started successfully after that. (Disappointingly, I'm still seeing a large banner about the CVE requiring an upgrade to 4.4+ despite now already being on 4.8.1, but that is also not the point of this issue.)

I'll keep this open until the next release (including the fix from https://github.com/RocketChat/Rocket.Chat/pull/26058) is made, or one of the maintainers can close it sooner if desired.

**Comment by jorge-figueroa** [comment]
> @jacobgkau I fixed this by adding `Rate_Limiter_Limit_RegisterUser` to `rocketchat_settings`.
> 
> ```js
> db.rocketchat_settings.insert({ 
>   "_id" : "Rate_Limiter_Limit_RegisterUser",
>   "createdAt" : ISODate("2022-07-01T05:32:48.791Z"),
>   "value" : 1,
>   "packageValue" : 1,
>   "valueSource" : "packageValue",
>   "secret" : false,
>   "enterprise" : false,
>   "i18nDescription" : "Rate_Limiter_Limit_RegisterUser_Description",
>   "autocomplete" : true,
>   "sorter" : 0,
>   "ts" : ISODate("2022-07-01T05:32:48.791Z"),
>   "type" : "int",
>   "group" : "Rate Limiter",
>   "section" : "Feature_Limiting",
>   "enableQuery" : "{\"_id\":\"API_Enable_Rate_Limiter\",\"value\":true}",
>   "i18nLabel" : "Rate_Limiter_Limit_RegisterUser",
>   "hidden" : false,
>   "blocked" : false,
>   "requiredOnWizard" : false,
>   "env" : false,
>   "public" : false,
>   "_updatedAt" : ISODate("2022-07-01T05:38:12.246Z")
> })
> ```

I have same issue. How do you apply this change?

**Comment by jacobgkau** [comment]
> I have same issue. How do you apply this change?

On an Ubuntu system running the snap, enter your Mongo CLI with `rocketchat-server.mongo`, then run `use parties` to switch to the correct database, then copy/paste the entire block in and press enter again.

**Comment by neylwalecki** [comment]
Thanks guys! I have update from 3.x to 4.x today because just like you and got the "Error 502 Bad Request" and the same "Rate_Limiter_Limit_RegisterUser" warning. The fix you posted worked like a charm! :-)

**Comment by skywidetech** [comment]
it looks like that 3.18.7 already fixed cve-2022-32211. https://github.com/RocketChat/Rocket.Chat/releases?page=2
Correct me if I am wrong. 
I have just downgraded back to 3.18.7 and wait for the next patch to fix the "indexes" and "numRequestsAllowed" problem

**Comment by skywidetech** [comment]
Hi guys, instead of jumping directly from 3.18 to 4.8.7. I tried to upgrade from 3.18 to 4.6.4 first (the second file on the releases page), and then I upgrade from 4.6.4 to 4.8.1 which looks like working for me. Try it out.

**Comment by wreiske** [comment]
https://github.com/RocketChat/Rocket.Chat/issues/24347#issuecomment-1172982771

**Comment by dudanogueira** [comment]
Hi!

This will be fixed in 5.0:
https://github.com/RocketChat/Rocket.Chat/pull/26058

**Comment by wilkis3** [comment]
Will be there an update in the snap channel soon?

**Comment by geekgonecrazy** [comment]
This was also released in 4.8.2

**Comment by jacobgkau** [comment]
4.8.3 and 5.x are both released to stable, so I'm closing this as completed.
